:art: Implement quality categories

aldebaran30701 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* Follow the check list items defined by https://github.com/constellation-app/constellation/blob/master/CONTRIBUTING.md#pull-requests
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix`, `hotfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Description of the Change
This PR adds Quality categories to replace the visual representation of quality scores. 
It does not completely rid the code of "scores", as completely getting rid of the abstract
`getQuality()` method would cause other rules to break.

The QualityControlView table and the Data access view will now show categories of quality, as opposed to scores.

The categories map to int values which are representative of the old values. for example, the quality score previously used will map to a category. 
* MissingTypeRule RISK = 90 , maps to SEVERE.

DEFAULT = 1 - 29
INFO = 30 - 59
WARNING = 60 - 89
SEVERE = 90 - 94
FATAL = 95 +

An alert box now is implemented to map rules to each priority level. Allowing for manual overriding of specific rules or rules which are important to the user.
![image](https://user-images.githubusercontent.com/58677759/89250960-0f91d080-d655-11ea-8abe-0d8b44002ffd.png)

Highest priority rule in violation gets that label added.
![image](https://user-images.githubusercontent.com/58677759/89251039-3c45e800-d655-11ea-8598-2872179a5fab.png)

Rule dialog shows in categories and also ranks highest priority category first.
![image](https://user-images.githubusercontent.com/58677759/89251071-4bc53100-d655-11ea-9770-1241716b9999.png)

Can map a rule to the default category, allowing that rule to be completely ignored.
![image](https://user-images.githubusercontent.com/58677759/89251104-5c75a700-d655-11ea-9b17-b57f8c350fde.png)

Highest priority category in action gets added to the DAV button
![image](https://user-images.githubusercontent.com/58677759/89251124-68f9ff80-d655-11ea-833e-47efa7769aa1.png)


<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
N/A
<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
Feature allows user to easily read the quality of the data
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an exising file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits
Feature allows user to easily read the quality of the data
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
None realised
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
loading different data, testing sort order, testing any place score was used previously.
opening up dialog boxes, switching priorities.
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
#378 
<!-- Link any applicable issues here -->

aldebaran30701 commented:
This is ready for review, it appears that the 2 sonar warnings are false positives. Unless the issue is staring me in the face!

arcturus2 commented on CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java:
```diff
@@ -70,38 +79,44 @@
 
-    public QualityControlViewPane(QualityControlViewTopComponent topComponent) {
-        this.parent = topComponent;
+    static {
+        for (final QualityControlRule rule : Lookup.getDefault().lookupAll(QualityControlRule.class)) {
+            rulePriorities.put(rule, rule.getCategory(0));
+        }
+    }
```
Running this in a static block like this means this bit of code must run when the class is loaded into memory (i.e when Constellation loads). This becomes problematic for a couple of reasons:
1) Constellation load times will take longer.
2) There is actually a runtime exception that Java throws when there are too many things run in a static block like this and a while back I moved them to be called lazily instead which you can do in `getPriorities()` with something like:

```java
private static Map<QualityControlRule, QualityCategory> rulePriorities = null;
...
...
public static Map<QualityControlRule, QualityCategory> getPriorities() {
    if(rulePriorities == null) {
        for (final QualityControlRule rule : Lookup.getDefault().lookupAll(QualityControlRule.class)) {
            rulePriorities.put(rule, rule.getCategory(0));
        }
    }
}

return rulePriorities;
```

FYI the call to `Lookup` isn't super efficient. For instance I have been able to dramatically shave seconds off the execution just by caching the object returned by `Lookup` where as it was being called many times (like in for loops in other parts of the app).

aldebaran30701 replied:
Hi @arcturus2 , Thanks for all the feedback! 👍 

>With the custom priorities radio option section, I don't believe this is being persisted into a Preferences object so it looks like changes will be lost after restarting Constellation.

You are correct, it doesn't currently.
I've been looking to store it into a preferences object but have hit a snag. I want to store a map into the preferences but have not found a method to set an object into the preferences. it only takes String or numeric values. 
From this I've pivoted to converting the map to a json String. I've looked through the utils and haven't found anything out the box. Do you have any suggestions re this or another possible solution?

>It didn't look like any of the "Rule" classes like UnknownTypeRule needed any changes; that's interesting and good actually.

Ideally this was what I was going for so I'm glad it all works without needing further extensions to those classes.


>Can you please write some unit tests to cover the logic? If you want a hand just let me know and happy to help writing uni tests.

>There should be a mention about this change to the users via the whatsnew.txt file informing them that scores are replaced by categories.

I can do these. 🍰 


>Running this in a static block like this means this bit of code must run when the class is loaded into memory (i.e when Constellation loads). 

Ohh no worries, I think that should be an easy fix.


>FYI the call to Lookup isn't super efficient. For instance I have been able to dramatically shave seconds off the execution just by caching the object returned by Lookup

I could store it into a class level variable, but when would i know to update it? only on instantiation of the view? 

I'll be a little while getting these changes in, and will re-request review once it's all wrapped up. 
Once again, cheers for taking such an in-depth look at this. 🙏

arcturus2 replied:
>...From this I've pivoted to converting the map to a json String. I've looked through the utils and haven't found anything out the box. Do you have any suggestions re this or another possible solution?

Yep that is how some other preferences work when you need to do more than a String. Have a look at your boostrap folder and then `build\testuserdir\config\Preferences\au\gov\asd\tac\constellation\plugins.properties`. Its a json string which `RecentParameterValues` loads and saves.

> I could store it into a class level variable, but when would i know to update it? only on instantiation of the view?
Yep that's how I would do it. You should be fine to put it in the class where `getPriorities` gets called.

aldebaran30701 commented:
Hi @arcturus2 , 
Just a heads up this is what I've changed.

 *Implemented Preferences object for mapping of categories to rules
 *Added JSON Utilities for writing a Map to String and vice versa.
 *Added entry into `whatsnew.txt`
 *Lazy initialisation over rulePriorities Map
 *Lazy initialisation and caching of lookup object

Added unit tests for:
 *JSON Utilities
 *QualityControlEvent
 *Quality Control Rule

Should be all of the suggestions 👍

arcturus2 commented:
Hi @aldebaran30701 so I got an NPE at line `QualityControlEvent.java:87`. I tried to continued testing by putting in 

```
category = QualityControlViewPane.getPriorities().get(rule) != null ?
                            QualityControlViewPane.getPriorities().get(rule):
                            QualityCategory.DEFAULT;
```

Next, for some reason I am not seeing your GUI changes like the category appearing either. I get this:

![Screenshot from 2020-08-07 09-47-47](https://user-images.githubusercontent.com/39325530/89593716-376e7780-d893-11ea-8a88-7caaaaa529d3.png)

aldebaran30701 commented:
>Hi @aldebaran30701 so I got an NPE at line QualityControlEvent.java:87.

Now that I read this, I'm thinking I missed the test case where the user does not have any preferences saved in memory. I'll verify all my logic and test this case to check if that's what's causing it. If that is the case, It would be why no GUI is updating. It defaults to Default category which is shown.

arcturus2 commented:
I see @aldebaran30701. Again this is great and I want to merge this is in as soon as possible but I would like one more thing done. If the requirement is that a user has to set a preference for each rule then that will be a problem if they don't.

So what we generally do is have a default implementation. This could be made via a `ServiceProvider` annotation (meaning you use Lookup) to provide a default implementation for the preference. 

This would also allow other module suites to override the default preference if they don't agree with what the default implementation provided. And the user could also override it too. The user overriding vs the ServiceProvider is a new concept though.

@cygnus-x-1 what is your thoughts on all of this and this PR in general given you wanted the user to be able to customise the severity as well.

aldebaran30701 commented:
@arcturus2 So I've fixed that bug, somewhere along the execution path it was initializing the array as empty, but not loading the look-ups. Now even if the array is empty, it re-checks the look-up.

>If the requirement is that a user has to set a preference for each rule then that will be a problem if they don't.

So straight out the box this Pr will use the default category linked with that rule. This is done via the rules "RISK" attribute.
If a preference is stored on that PC, then it uses the category defined by that rules preference.

>So what we generally do is have a default implementation

What do you mean by this? From my understanding my above comment is what a default implementation would mean. Let me know if I am reading it wrong 😆 

It could benefit from a "reset to default" button which allowed the user to use the default rule categories. Let me know if you agree.

arcturus2 commented:
> > So what we generally do is have a default implementation
> 
> What do you mean by this? From my understanding my above comment is what a default implementation would mean. Let me know if I am reading it wrong 😆

All good. I was actually thinking you did this via the score but I went off on a tangent. What you have done is perfect. I will do more testing but all good.

arcturus2 commented:
> It could benefit from a "reset to default" button which allowed the user to use the default rule categories. Let me know if you agree.

Based on testing and modifying my selection i think having a reset button would be extremely useful. I think debugging problems with users would be much easier by asking them to "reset to default" and try again first.

With regards to the placement of the button I'd vote to have it inside the dialog where you can customise the severity.

Also a few minor things.

1.On Linux the dialog window is a little smaller so the categories are elipsed. Could you look at extending the width if its not too much trouble?
2. Can you set a title for the dialog as it defaults to "Message".
3. Can we change the wording a bit from "Priority defines category levels for each rule" to "Customise the priority of rules" or something like that?

![image](https://user-images.githubusercontent.com/39325530/89743577-d095cc00-dae7-11ea-81ec-54ee3329b55f.png)

aldebaran30701 commented:
I've gave the dialog a bit of extra space for other configurations. How does it look? 
![image](https://user-images.githubusercontent.com/58677759/89744738-fa9ebc80-daee-11ea-8490-333c5230bda6.png)

arcturus2 commented:
Looks much better. Thank you @aldebaran30701 

![image](https://user-images.githubusercontent.com/39325530/89745558-fd51df80-daf7-11ea-8ce4-09364d20b224.png)

mimosa2 commented:
Good work @aldebaran30701. From my testing it looks very good. Some minor fixes can be as follows;
1. The preference file entry (see below) can have a label such as "customRules=" . This is similar to recentValues
  ={"Missing type":"WARNING","Unknown type":"DEFAULT","Identifier inconsistent with type":"INFO"}
2. The alert box w/o any rules can have message such as 'There is no rules for this Identifier"

![Rules_Alert](https://user-images.githubusercontent.com/54261086/89748815-7bb67d80-db08-11ea-9a11-4c9a806fe530.PNG)

aldebaran30701 commented:
> 1. The preference file entry (see below) can have a label such as "customRules=" . This is similar to recentValues
>    ={"Missing type":"WARNING","Unknown type":"DEFAULT","Identifier inconsistent with type":"INFO"}
> 2. The alert box w/o any rules can have message such as 'There is no rules for this Identifier"

Much appreciated for the feedback @mimosa2 
I've addressed these in a different ticket and PR as this has been since merged.  Please see #769 for the ticket and #770 for the PR.
